install.sh

```diff
@@ -312,6 +324,64 @@ mkdir -p "$DIR"
 cd "$DIR"
 
 # ========================================================================================
+# Sandcats.io - Dynamic DNS service hosted by Sandstorm.io
+if [ "yes" = "$USE_SANDCATS" ] ; then
+  # We generate the public key before prompting for a desired hostname
+  # so that when the user presses enter, we can try to register the
+  # hostname, and if that succeeds, we are totally done. This avoids a
+  # possible time-of-check-time-of-use race.
+  echo "As a Sandstorm user, you are invited to use a free Internet hostname as a subdomain of sandcats.io."
+
+  # The sandcats service places its authentication files in $DIR/sandcats.
+  if [ ! -f sandcats/id_rsa.private_combined ] ; then
+
+    # The openssl key generation process can take a few seconds, so we
+    # print a ... while that happens.
+    echo -n '...'
+
+    # We are already in $DIR. It's important to make it mode 0700
+    # because we store TLS client authentication keys here.
+    mkdir -p -m 0700 sandcats
+    chmod 0700 sandcats
+
+    # Generate key for client certificate. OpenSSL will read from
+    # /dev/urandom by default, so this won't block. We abuse the ``
+    # operator so we can have inline comments in a multi-line command.
+    openssl \
+      req `# Invoke OpenSSL's PKCS#10 X.509 bits.` \
+      -new `# Create a new certificate/request.` \
+      -newkey rsa:4096 `# Create a new RSA key of length 4096 bits.` \
+      -days 3650 `# Make the self-signed cert valid for 10 years.` \
+      -nodes `# no DES -- that is, do not encrypt the key at rest.` \
+      -x509 `# Output a certificate, rather than a signing request.` \
+      `# Sandcats ignores the subject in the certificate; use` \
+      `# OpenSSL defaults.` \
+      -subj "/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd" \
+      -keyout sandcats/id_rsa `# Store the resulting RSA private key in id_rsa` \
+      -out sandcats/id_rsa.pub `# Store the resulting certificate in id_rsa.pub` \
+      2>/dev/null `# Silence the progress output.`
+
+    # We combine these two things into one glorious all-inclusive file
+    # for the `curl` command. It is just as private as id_rsa.
+    cat sandcats/id_rsa sandcats/id_rsa.pub > sandcats/id_rsa.private_combined
+
+    # Set filesystem permissions, in case the files get copied into the
+    # wrong place later.
+    chmod 0600 sandcats/id_rsa sandcats/id_rsa.pub sandcats/id_rsa.private_combined
+
+    # Delete the "..." that we left on the screen, to indicate that we
+    # are done generating cryptographic material.
+    echo -ne '\b\b\b'
```

Comment:
Not important, but FWIW this doesn't actually delete the characters, just moves the cursor back. You have to additionally print spaces to overwrite the chars. On another note, you could use `\r` to mean "go back to beginning of line".